In-app users admin

zomars opened the pull request:
## What does this PR do?

- Adds a new enterprise-only feature for ADMINs
- Allows ADMIN users to manage their instance users
- Adds UI to add users
- Adds UI to edit basic user info
- Adds UI to list users and delete from there

## Type of change

<!-- Please delete bullets that are not relevant. -->

- New feature (non-breaking change which adds functionality)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [ ] As and ADMIN go to `/settings/admin/users`
- [ ] You should see your current users
- [ ] Try to add a new user
- [ ] Try to edit a current user
- [ ] Try to delete a current user

TODO in a follow-up

- Add search functionality user listing
- Add pagination for user listing
- Add more editable fields
  - metadata (for billing issues)
  - passwords?
  - Allow uploading/editing  avatars
- Add impersonation quick link?

Tagging @Jaibles and @keithwillcode for RFC

## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->

- I haven't performed a self-review of my own code and corrected any misspellings
- I haven't added tests that prove my fix is effective or that my feature works

zomars committed:
WIP

zomars committed:
WIP

zomars committed:
Merge branch 'main' into feat/users

zomars committed:
Update trpc-router.ts

zomars committed:
Fixes

zomars committed:
Update trpc-router.ts

zomars commented on packages/features/ee/users/schemas/userBodySchema.ts:
```diff
@@ -0,0 +1,12 @@
+
+import { optionToValueSchema } from "@calcom/prisma/zod-utils";
+
+export const userBodySchema = z
```
Used to transform Select values to plain ones

zomars commented on packages/features/settings/layouts/SettingsLayout.tsx:
```diff
@@ -87,7 +88,7 @@ const tabs: VerticalTabItemProps[] = [
       { name: "impersonation", href: "/settings/admin/impersonation" },
       { name: "apps", href: "/settings/admin/apps/calendar" },
-      { name: "users", href: "https://console.cal.com" },
+      { name: "users", href: "/settings/admin/users" },
```
Since console doesn't work currently

zomars commented on packages/features/settings/layouts/SettingsLayout.tsx:
```diff
@@ -414,7 +415,9 @@ export default function SettingsLayout({
           <ShellHeader />
-          <ErrorBoundary>{children}</ErrorBoundary>
+          <ErrorBoundary>
+            <Suspense fallback={<FiLoader />}>{children}</Suspense>
```
So we can use Suspense in settings. It is ignored if suspense isn't used.

zomars committed:
Update zod-utils.ts

zomars commented on packages/ui/components/table/Table.tsx:
Ported from console

zomars reviewed (commented):
Ready for review

PeerRich commented:
would it be possible to also change identity providers from google to email + password?

zomars committed:
Merge branch 'main' into feat/users

zomars committed:
Merge branch 'main' into feat/users

leog commented on packages/features/ee/users/components/UserForm.tsx:
```diff
@@ -0,0 +1,228 @@
+  ];
+
+  const userRoleOptions = [
+    { value: "USER", label: "USER" },
```
```suggestion
    { value: "USER", label: t("user") },
```

leog reviewed (approved):
Code looks good, and works locally as expected. Suggested a nit for labels.

Other notes that could be tackled later or dismissed if not relevant:
  1. I kept getting the following error for some strange reason, but everything worked anyway
  <kbd><img src="https://user-images.githubusercontent.com/467258/231507250-7d19d96a-6553-44f3-92ba-daa42a8c325c.png" /></kbd>
  2. Title when updating a user shows the user ID
  <kbd><img width="200" src="https://user-images.githubusercontent.com/467258/231507571-e0b0e918-3ade-4548-9d86-7fb5b90dcc76.png" /></kbd>
  3. When creating a new user, button reads "Update" in all caps, should be "Create" or "Save"
  4. Avatar was not being saved, I kept seeing the gravatar one.

zomars committed:
Merge branch 'main' into feat/users

zomars committed:
Apply suggestions from code review

Co-authored-by: Leo Giovanetti <hello@leog.me>

zomars committed:
Feedback and dark mode fixes

zomars committed:
Merge branch 'main' into feat/users

zomars committed:
Feedback

zomars committed:
Fixes

zomars commented:
> I kept getting the following error for some strange reason, but everything worked anyway

It might have something to do with experimental Suspense, fixed now.

> Title when updating a user shows the user ID

Changes to username

> When creating a new user, button reads "Update" in all caps, should be "Create" or "Save"

Fixed

> Avatar was not being saved, I kept seeing the gravatar one.

Removed for now, will fix in a follow up.

zomars committed:
Update UserForm.tsx
